pkg/webhook/overcommit/handler.go

```diff
@@ -0,0 +1,207 @@
+		http.Error(w, fmt.Sprintf("cannot encode response: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if _, err := w.Write(resp); err != nil {
+		klog.Errorf("Cannot write response: %v", err)
+		http.Error(w, fmt.Sprintf("cannot write response: %v", err), http.StatusInternalServerError)
+		return
+	}
+}
+
+func Mutate(ctx context.Context, data []byte) *v1.AdmissionResponse {
+	response := &v1.AdmissionResponse{}
+	// allow by default
+	response.Allowed = true
+
+	klog.Infof("Body: %s", strings.ReplaceAll(string(data), "\r", ""))
+	ar := v1.AdmissionReview{}
+	if err := json.Unmarshal(data, &ar); err != nil {
+		klog.Error(err)
+		response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+		return response
+	}
+
+	response.UID = ar.Request.UID
+
+	var pod corev1.Pod
+	if err := json.Unmarshal(ar.Request.Object.Raw, &pod); err != nil {
+		klog.Errorf("Cannot unmarshal raw object: %v", err)
+		klog.Errorf("Object: %v", strings.ReplaceAll(string(ar.Request.Object.Raw), "\r", ""))
+		response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+
+		return response
+	}
+
+	cfg, err := config.GetConfig()
+	if err != nil {
+		klog.Errorf("Cannot load config: %v", err)
+		response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+		return response
+	}
+
+	cs, _ := versioned.NewForConfig(cfg)
+	overCommit, err := cs.ArbiterV1alpha1().OverCommits().Get(ctx, OvercommitName, metav1.GetOptions{})
+
+	if err != nil {
+		klog.Errorf("Get overcommit resource error: %v", err)
+		response.Result = &metav1.Status{
+			Message: err.Error(),
+		}
+		return response
+	}
+
+	patch := PodPatch(&pod, overCommit)
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		klog.Errorf("Error marshaling pod update: %v", err)
+		return &v1.AdmissionResponse{
+			Result: &metav1.Status{
+				Message: err.Error(),
+			},
+		}
+	}
+
+	response.Patch = patchBytes
+	response.PatchType = func() *v1.PatchType { pt := v1.PatchTypeJSONPatch; return &pt }()
+
+	return response
+}
+
+func PodPatch(pod *corev1.Pod, overCommit *v1alpha1.OverCommit) []PatchEntry {
+	limitCPU := resource.MustParse(overCommit.Spec.CPU.Limit)
+	limitMemory := resource.MustParse(overCommit.Spec.Memory.Limit)
+	ratioCPU, _ := strconv.ParseFloat(overCommit.Spec.CPU.Ratio, 64)
+	ratioMemory, _ := strconv.ParseFloat(overCommit.Spec.Memory.Ratio, 64)
+
+	requestCPU := fmt.Sprintf("%vm", int64(float64(limitCPU.MilliValue())*ratioCPU))
+	requestMemory := fmt.Sprintf("%vKi", int64(float64(limitMemory.Value())*ratioMemory)/1024)
+
+	initContainers := make([]corev1.Container, 0)
+	for _, c := range pod.Spec.InitContainers {
+		c.Resources = corev1.ResourceRequirements{
+			Limits: corev1.ResourceList{
+				"cpu":    limitCPU,
+				"memory": limitMemory,
+			},
+			Requests: corev1.ResourceList{
+				"cpu":    resource.MustParse(requestCPU),
+				"memory": resource.MustParse(requestMemory),
+			},
+		}
+		initContainers = append(initContainers, c)
```

Comment:
We should not assume that the request and limit in the spec of the user's pod have only 2 options: cpu or mem, and that the current code will ignore the other resources, which is incorrect.